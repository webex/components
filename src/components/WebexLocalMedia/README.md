# Webex Local Media Component

Webex local media component displays the user local stream, if enabled. If user's local media is disabled the avatar of the current _Webex_ user is displayed.

<p align="center">
  <img src="./WebexLocalMedia.png" alt="Default Webex Local Media" />
</p>

## Preview

To see all the different possible states of the Webex Local Media component,
you can view our [storybook](https://webex.github.io/components/storybook/?path=/story/webex-local-media--enabled)
or run the following **NPM** command:

```shell
  npm start
```

## Embed

1. Create a component adapter from which the data will be retrieved (See [adapters](../../adapters)). For instance:

    ```js
    const jsonAdapter = new WebexJSONAdapter(jsonData);
    ```

2. Create a component instance by passing the person ID as a string and enclose it within
[a data provider](../WebexDataProvider/WebexDataProvider.js) that takes
the [component data adapter](../../adapters/WebexJSONAdapter.js) that we created previously

    ```js
    <WebexDataProvider adapter={jsonAdapter}>
      <WebexLocalMedia meetingID="meetingID" />
    </WebexDataProvider>
    ```

The component knows how to manage its data. If anything changes in the data source that the adapter manages,
the component will also update on its own.
