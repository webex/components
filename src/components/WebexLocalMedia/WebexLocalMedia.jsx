import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {Spinner} from '@momentum-ui/react';

import WebexAvatar from '../WebexAvatar/WebexAvatar';
import {WEBEX_COMPONENTS_CLASS_PREFIX} from '../../constants';
import {PHONE_LARGE} from '../breakpoints';
import {
  useElementDimensions,
  useMe,
  useMeeting,
  useStream,
} from '../hooks';

/**
 * Webex Local Media component displays the user's local video or local share.
 *
 * @param {object} props  Data passed to the component
 * @param {string} props.meetingID  ID of the meeting from which to obtain local media
 * @param {string} props.mediaType  Type of local media to display
 * @param {string} props.className  Custom CSS class to apply
 * @returns {object} JSX of the component
 */
export default function WebexLocalMedia({className, meetingID, mediaType}) {
  const [mediaRef, {width}] = useElementDimensions();
  const {localVideo, localShare} = useMeeting(meetingID);
  let stream;

  switch (mediaType) {
    case 'video':
      stream = localVideo;
      break;
    case 'screen':
      stream = localShare;
      break;
    default:
      break;
  }

  const ref = useStream(stream);
  const {ID} = useMe();

  const classBaseName = `${WEBEX_COMPONENTS_CLASS_PREFIX}-local-media`;
  const mainClasses = {
    [classBaseName]: true,
    [`${classBaseName}-desktop`]: width >= PHONE_LARGE,
    [`${classBaseName}-no-media`]: !stream,
    [className]: !!className,
  };

  const disabledVideo = ID ? <WebexAvatar personID={ID} displayStatus={false} /> : <Spinner />;

  return (
    <div ref={mediaRef} className={classNames(mainClasses)}>
      {
        /* eslint-disable-next-line jsx-a11y/media-has-caption */
        stream ? <video ref={ref} playsInline autoPlay /> : disabledVideo
      }
    </div>
  );
}

WebexLocalMedia.propTypes = {
  className: PropTypes.string,
  meetingID: PropTypes.string.isRequired,
  mediaType: PropTypes.oneOf(['video', 'screen']),
};

WebexLocalMedia.defaultProps = {
  className: '',
  mediaType: 'video',
};
