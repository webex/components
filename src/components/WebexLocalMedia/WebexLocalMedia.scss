.local-media {
  display: flex;
  justify-content: center;
  align-items: center;

  height: 100%;
  width: 100%;
  min-height: 7.5rem; // Smallest height for local media is the tallest an avatar can get through transforms
  min-width: 7.5rem; // Smallest width for local media is the widest an avatar can get through transforms

  background: black;

  video {
    height: 100%;
    width: 100%;
  }

  .md-avatar {
    transform: scale(2); // Avatar is 40x40px, this makes it 80x80px
    transform-origin: center center;
  }

  &-desktop {
    .md-avatar {
      transform: scale(3); // Avatar is 40x40px, this makes it 120x120px
    }
  }
}

.no-media {
  background: white;
}
