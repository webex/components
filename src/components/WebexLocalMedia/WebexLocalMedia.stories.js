import React from 'react';
import {storiesOf} from '@storybook/react';

import jsonData from '../../data';
import {WebexJSONAdapter} from '../../adapters';
import {WebexLocalMedia, WebexDataProvider} from '../';

const stories = storiesOf('Webex Local Media', module);
const webexAdapter = new WebexJSONAdapter(jsonData);

stories.add('enabled', () => (
  <WebexDataProvider adapter={webexAdapter}>
    <WebexLocalMedia meetingID="localVideo" />
  </WebexDataProvider>
));

stories.add('disabled', () => (
  <WebexDataProvider adapter={webexAdapter}>
    <WebexLocalMedia meetingID="noMedia" />
  </WebexDataProvider>
));
