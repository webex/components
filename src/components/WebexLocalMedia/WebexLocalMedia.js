import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {Spinner} from '@momentum-ui/react';

import {WEBEX_COMPONENTS_CLASS_PREFIX} from '../../constants';
import {WebexAvatar} from '../';
import {useElementDimensions, useMe, useMeeting, useStream} from '../hooks';
import {PHONE_LARGE} from '../breakpoints';

import './WebexLocalMedia.scss';

/**
 * Webex Local Media component displays the user's local video.
 *
 * @param {string} props.meetingID  ID of the meeting from which to obtain local media
 * @returns {Object} JSX of the component
 */
export default function WebexLocalMedia({className, meetingID}) {
  const [mediaRef, {width}] = useElementDimensions();
  const {localVideo} = useMeeting(meetingID);
  const {ID} = useMe();
  const videoRef = useStream(localVideo);

  const classBaseName = `${WEBEX_COMPONENTS_CLASS_PREFIX}-local-media`;
  const mainClasses = {
    [classBaseName]: true,
    [`${classBaseName}-desktop`]: width >= PHONE_LARGE,
    [`${classBaseName}-no-media`]: localVideo === null,
    [className]: !!className,
  };
  const disabledVideo = ID ? <WebexAvatar personID={ID} displayStatus={false} /> : <Spinner />;

  return (
    <div ref={mediaRef} className={classNames(mainClasses)}>
      {localVideo ? <video ref={videoRef} playsInline autoPlay /> : disabledVideo}
    </div>
  );
}

WebexLocalMedia.propTypes = {
  className: PropTypes.string,
  meetingID: PropTypes.string.isRequired,
};

WebexLocalMedia.defaultProps = {
  className: '',
};
