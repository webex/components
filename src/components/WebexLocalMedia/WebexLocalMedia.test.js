import React from 'react';

import WebexLocalMedia from './WebexLocalMedia';

jest.mock('../hooks/useMe');
jest.mock('../hooks/useMeeting');
jest.mock('../hooks/useStream');

describe('Webex Local Media component', () => {
  describe('snapshot', () => {
    test('matches snapshot of local video', () => {
      expect(shallow(<WebexLocalMedia meetingID="localVideo" />)).toMatchSnapshot();
    });

    test('matches snapshot of disabled local video', () => {
      expect(shallow(<WebexLocalMedia meetingID="noMedia" />)).toMatchSnapshot();
    });

    test('matches snapshot of local video with custom CSS class', () => {
      expect(shallow(<WebexLocalMedia className="my-custom-class" meetingID="noMedia" />)).toMatchSnapshot();
    });
  });
});
