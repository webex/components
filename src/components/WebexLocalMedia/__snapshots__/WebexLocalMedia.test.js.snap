// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webex Local Media component snapshot matches snapshot of disabled local video 1`] = `
<div
  className="wxc-local-media wxc-local-media-no-media"
>
  <WebexAvatar
    displayStatus={false}
    personID="default"
  />
</div>
`;

exports[`Webex Local Media component snapshot matches snapshot of local video 1`] = `
<div
  className="wxc-local-media"
>
  <video
    autoPlay={true}
    playsInline={true}
  />
</div>
`;

exports[`Webex Local Media component snapshot matches snapshot of local video with custom CSS class 1`] = `
<div
  className="wxc-local-media wxc-local-media-no-media my-custom-class"
>
  <WebexAvatar
    displayStatus={false}
    personID="default"
  />
</div>
`;
